### ISSUE TITLE
Issues with packaging

### ISSUE THREAD

**Issue by mp-pinheiro** [rdeioris/LuaMachine#31]
Hey! How's it going? Hope y'all are doing good.

First of all, as always, I want to thank you for the fantastic plugin! I work on a small indie game studio and we've bought it on the Marketplace a couple of years ago and have been using it extensively ever since.

We're currently having some issues with packaging the game with the Lua scripts. Our project structure looks something like this:"

```
├── Content
├── ├── Lua
├── ├── ├── Modules
├── ├── ├── ├── module1.lua
├── ├── ├── ├── module2.lua
├── ├── ├── ├── ...
├── ├── ├── Controllers
├── ├── ├── ├── controller1.lua
├── ├── ├── ├── controller2.lua
├── ├── ├── ├── ...
├── ├── ├── main.lua
├── ├── ...
├── ...
├── Project.uproject
```
The `main.lua` script requires and uses the modules and controllers exemplified (i.e `module1.lua`, `controller1.lua`, etc), and is directly called in the UE (Unreal Engine) project by using the `Lua Run File` blueprint node.

![image](https://user-images.githubusercontent.com/17935077/127577678-016cf0f5-8b77-41e4-90d9-38012053e0ff.png)

The `main.lua` file is not the only script directly called by UE, there are other calls like this one:

![image](https://user-images.githubusercontent.com/17935077/127578345-5b8d5585-f2c6-4b86-aa94-0c98694c655d.png)

Everything works perfectly when running from within the engine, and it allows for easily editing the scripts without having to reimport them every time, which we strive as Lua users.

However, we can't seem to figure out how to make packaging work with this configuration. We've tried following the specifications on the Readme page by adding the `Lua` directory as a non-asset directory to be packaged:

![image](https://user-images.githubusercontent.com/17935077/127577845-b6278474-fc7c-4533-919f-c450b229fb30.png)

Unfortunately, this doesn't work, as the packaged game is unable to find the Lua files, with all `Lua Run File` calls failing. 

Manually copying our Lua folder and files to the newly packaged game Content folder does work, but this leaves our Lua files very much exposed, which would be less than ideal.

Could you guys help point us in the right direction?

Another thing, and I'm not sure if it's relevant or not, but in order for the `Lua Run File` calls to work, our `Lua State` blueprint has the following config:

![image](https://user-images.githubusercontent.com/17935077/127577556-c835ebf9-edd1-48c6-83db-dc1ca8efc3a3.png)

If the `Lua` directory is removed from the config, the `Lua Run File` calls fail saying the file was not found.

**Comment by rdeioris** [comment]
Hi, your procedure looks correct. A random list of stuff to check:

* use the UnrealPak.exe commands (it is in the Epic Engine/Binaries/Win64 directory) to check if the Lua scripts are correctly included in the pakfile (they must be under Content/Lua/)
* The Append Project Content Sub Dir is for adding search paths to the the lua package resolution procedure (so you can use require easily), can you check if by removing it you can still run the Lua/main.lua file ?
* (Just as a test) can you use the 'Additional Non-Asset Drectories to copy' to check if the system still works ?

**Comment by mp-pinheiro** [comment]
Hey, thanks for the swift response! I'll from easiest to hardest to answer:

> (Just as a test) can you use the 'Additional Non-Asset Drectories to copy' to check if the system still works ?

It works fine when this is set, the source Lua files are copied to the Packaged directory and the game runs as intended.

> The Append Project Content Sub Dir is for adding search paths to the the lua package resolution procedure (so you can use require easily), can you check if by removing it you can still run the Lua/main.lua file

If I remove `Lua` from the `Append Project Content Sub Dir` option, everything stops working. The `main.lua` is found and runs, but the other lua files that it includes are not:

```
LogLuaMachine: Error: Lua execution error: ...rojects/UEProject/Content//Lua\Controllers\controller1.lua:27: module 'Module1' not found:
    no field package.preload['Module1']
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\lua\Module1.lua'
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\lua\Module1\init.lua'
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\Module1.lua'
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\Module1\init.lua'
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\..\share\lua\5.3\Module1.lua'
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\..\share\lua\5.3\Module1\controller1.lua'
    no file '.\Module1.lua'
    no file '.\Module1\controller1.lua'
    no file 'D:/Documents/Unreal Projects/UEProject/Content//Module1.lua'
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\Module1.dll'
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\..\lib\lua\5.3\Module1.dll'
    no file 'C:\Program Files\Epic Games\UE_4.25\Engine\Binaries\Win64\loadall.dll'
    no file '.\Module1.dll'
```

> use the UnrealPak.exe commands (it is in the Epic Engine/Binaries/Win64 directory) to check if the Lua scripts are correctly included in the pakfile (they must be under Content/Lua/)

This one is the most confusing to me. I ran `UnrealPak UEProject-WindowsNoEditor.pak -list`, but got the follownig output:

```
LogPakFile: Display: Using command line for crypto configuration
LogPakFile: Display: Mount point /
LogPakFile: Display: 0 files (0 bytes), (0 filtered bytes).
LogPakFile: Display: Unreal pak executed in 0.000566 seconds
```

If I understand that output correctly, it looks like nothing was packed into the pak file, which makes no sense. I also tried extracting the pak file to a directory with `UnrealPak UEProject-WindowsNoEditor.pak -extract "out"`, but to no avail. This is very strange considering the pak file is 800mb in size, and, as mentioned before, the game runs perfectly if I copy the Lua folder.

**Comment by rdeioris** [comment]
@mp-pinheiro so i took a bit of time for digging into the issue, and at the end i realized that a better lua module loaders for packaged games was required. It is way smarter than the one before (it can even load code asset with require('Game/YourAsset') !), but more important it honours the search setup even for packaged assets. Just pull from the master and let me know how it works for you. Regarding the unrealpak tool not showing files, well it is a bit of a mistery, but i think we can ignore it :)

**Comment by mp-pinheiro** [comment]
> @mp-pinheiro so i took a bit of time for digging into the issue, and at the end i realized that a better lua module loader for packaged games was required. It is way smarter than the one before (it can even load code asset with require('Game/YourAsset') !), but more important it honours the search setup even for packaged assets. Just pull from the master and let me know how it works for you. Regarding the unrealpak tool not showing files, well it is a bit of a mistery, but i think we can ignore it :)

Hey @rdeioris, that sounds fantastic! On our end, we figured out why our requires were failing when the `Append Project Content Sub Dir` was not set. We also have an overloaded `require` function to ease requirements. We made some changes and now the project runs fine wihouth the `Append Project Content Sub Dir` set to `Lua`.

I pulled from master and the project runs fine within the engine, but unfortunately our packaging is now failing to cook. The UE error logs are a mess, but this is what I could gather:

- This error LuaMachine error occurred repeatedly 
  ```
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestOwnage] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestOwnage.TestOwnage
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestMultiArgs] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestMultiArgs.TestMultiArgs
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorUFunctionAsValueRaw] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorUFunctionAsValueRaw.TestActorUFunctionAsValueRaw
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorUFunctionAsValue] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorUFunctionAsValue.TestActorUFunctionAsValue
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorUFunction] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorUFunction.TestActorUFunction
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorToLuaComponentFromState] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorToLuaComponentFromState.TestActorToLuaComponentFromState
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorToLuaComponentByStateAndName] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorToLuaComponentByStateAndName.TestActorToLuaComponentByStateAndName
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorToLuaComponentByName2] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorToLuaComponentByName2.TestActorToLuaComponentByName2
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorToLuaComponentByName] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorToLuaComponentByName.TestActorToLuaComponentByName
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorToLuaComponent] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorToLuaComponent.TestActorToLuaComponent
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorSetFieldOfTable] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorSetFieldOfTable.TestActorSetFieldOfTable
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorSetField] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorSetField.TestActorSetField
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorMethodValueMultiNoArgs] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorMethodValueMultiNoArgs.TestActorMethodValueMultiNoArgs
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorMethodMultiNoArgs] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorMethodMultiNoArgs.TestActorMethodMultiNoArgs
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorMethodMulti] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorMethodMulti.TestActorMethodMulti
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorMethod] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorMethod.TestActorMethod
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorInception] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorInception.TestActorInception
  ProcessResult.StdOut:   LogInit: Display: LogBlueprint: Error: [Compiler TestActorField] Spawn node  SpawnActor NONE  must have a  Class  specified. from Source: /LuaMachine/Tests/TestActorField.TestActorField    
  ```

  It happened on multiple test files (
- This is the final error message:
  ```
  Log.WriteException: CommandletException: Editor terminated with exit code 1 while running Cook for D:\Documents\Unreal Projects\UEProject\UEProject.uproject; see log C:\Users\mp-pi\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.25\Cook-2021.08.01-13.15.11.txt
  Log.WriteException:    at AutomationTool.CommandUtils.RunCommandlet(FileReference ProjectName, String UE4Exe, String Commandlet, String Parameters, String& DestLogFile) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\AutomationUtils\CommandletUtils.cs:line 424
  Log.WriteException:    at AutomationTool.CommandUtils.RunCommandlet(FileReference ProjectName, String UE4Exe, String Commandlet, String Parameters) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\AutomationUtils\CommandletUtils.cs:line 237
  Log.WriteException:    at AutomationTool.CommandUtils.CookCommandlet(FileReference ProjectName, String UE4Exe, String[] Maps, String[] Dirs, String InternationalizationPreset, String[] CulturesToCook, String TargetPlatform, String Parameters) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\AutomationUtils\CommandletUtils.cs:line 89
  Log.WriteException:    at Project.Cook(ProjectParams Params) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\Scripts\CookCommand.Automation.cs:line 247Wrapped by AutomationException: Cook failed.
  Log.WriteException:    at Project.Cook(ProjectParams Params) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\Scripts\CookCommand.Automation.cs:line 259
  Log.WriteException:    at BuildCookRun.DoBuildCookRun(ProjectParams Params) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\Scripts\BuildCookRun.Automation.cs:line 199
  Log.WriteException:    at BuildCookRun.ExecuteBuild() in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\Scripts\BuildCookRun.Automation.cs:line 39
  Log.WriteException:    at AutomationTool.BuildCommand.Execute() in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\AutomationUtils\BuildCommand.cs:line 263
  Log.WriteException:    at AutomationTool.Automation.Execute(List`1 CommandsToExecute, Dictionary`2 Commands) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\AutomationUtils\Automation.cs:line 547
  Log.WriteException:    at AutomationTool.Automation.Process(String[] Arguments, StartupTraceListener StartupListener) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\AutomationUtils\Automation.cs:line 511
  Log.WriteException:    at AutomationTool.Program.MainProc(String[] Arguments, StartupTraceListener StartupListener) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\Program.cs:line 176
  Log.WriteException:    at AutomationTool.Program.<>c__DisplayClass1_0.<Main>b__2() in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\Program.cs:line 88
  Log.WriteException:    at AutomationTool.InternalUtils.RunSingleInstance(Func`1 Main) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\AutomationUtils\Utils.cs:line 731
  Log.WriteException:    at AutomationTool.Program.Main(String[] Arguments) in D:\Build\++UE4\Sync\Engine\Source\Programs\AutomationTool\Program.cs:line 88
  Log.WriteException: ==============================================================================
  Program.Main: AutomationTool exiting with ExitCode=25 (Error_UnknownCookFailure)
  ```
 
Just to make sure I did everything correctly, I cloned the repo into a `Plugins/LuaMachine` directory and reopened the project for it to compile. We've been using the Marketplace version of the plugin for a while, so I'm not sure if that could impact anything.

**Comment by mp-pinheiro** [comment]
Just as a test, I set the repo to the commit right before the last (`96ecd68ca2f1ed241746011c5371a45519482dbc`) and the same error persists, so it doesn't seem related to the module loader changes.

**Comment by rdeioris** [comment]
@mp-pinheiro I strongly suggest you to build from visual studio (after having cloned the repository into Plugins and clicked on 'regenerate visual studio solution'). From the logs it looks like the plugin has not been built.

**Comment by mp-pinheiro** [comment]
Hi, sorry for the radio silence, we've been digging deep into this trying to figure out why our requires weren't working. I'll address everything by parts.

Regarding the build issue, I'm still unable to package the game when using the plugin cloning from master. Here is what I did:

1. Cloned the repository into `UEProject/Plugins`, which resulted in a `UEProject/Plugins/LuaMachine` directory containing all the repo files from master
1. Opened `UEProject.uproject`, which triggered a message saying I had missing modules and asking if I wanted to rebuild them, to which I said yes
    ![image](https://user-images.githubusercontent.com/17935077/128430026-00f27fb5-ae40-4b16-82fd-0d765396f55e.png)
1. Everything ran fine and the engine opened into our default map. I then went to `File -> Refresh Visual Studio Project`, which ran fine as well
    ![image](https://user-images.githubusercontent.com/17935077/128430603-9673b728-1884-4c41-b744-b2d6c519a516.png)
1. I then went to `File -> Open Visual Studio Project` and ran a build, which also ran fine
    ![image](https://user-images.githubusercontent.com/17935077/128430762-a9418d7d-a647-4933-81e2-5b569934aba8.png)
    ```
    1>Creating makefile for PrimateriaUEEditor (no existing makefile)
    1>Building PrimateriaUEEditor...
    1>Using Visual Studio 2019 14.28.29915 toolchain (C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910) and Windows 10.0.18362.0 SDK (C:\Program Files (x86)\Windows Kits\10).
    1>Building 1 action with 24 processes...
    1>  [1/1] PrimateriaUEEditor.target
    1>Total time in Parallel executor: 0.17 seconds
    1>Total execution time: 1.62 seconds
    ========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
    ```
1. Finally, I did our default packaging option: `File -> Package Project -> Windows 64`, but unfortunately, ran into the same cooking errors described in my last message. 

I'm not sure what I'm doing wrong, or if I'm skipping any steps.

---

I did manage to build the plugin, however. What I noticed on the error message was that all problems were related to Test files. Although this is far from ideal, I removed all `uasset` files that had `Test` in their names from the plugin. Doing that allowed the project to be packaged without any errors, so maybe there's some testing configs I'm missing.

With that done, we were able to test your changes, but our project, although working perfectly inside the engine, did not work when packaged. After a couple of long days debugging an testing things, we found out the root cause of the problem. We have a lot of overloaded global functions and methods to allow for a better usage of basic Lua features such as coroutines and package loading. We also have a [custom OOP class](https://github.com/yfrit/yfritlib/blob/master/Class.lua) that is used throughout our entire project, amongst other things.

Our package loading works something along these lines:

```lua
local oldRequire = _G.require

--create a loader that tries to require a Lua file everytime something is required
local function isModuleAvailable(name)
    local ok, errorMessage = pcall(oldRequire, name)
    if ok or not errorMessage:match(string.format("module '%s' not found", name)) then
        return true
    else
        return false
    end
end

--"Lua"
local attempts = {
    "Lua",
    "Lua/CardGameLogic",
    "Lua/MapExplorerLogic"
}

function _G.require(originalPath)
    originalPath = originalPath:gsub("%.", "/")

    for _, folderName in ipairs(attempts) do
        --"Module" => "Lua/Module" or "Lua/CardGameLogic/Module" or "Lua/MapExplorerLogic/Module"
        local newPath = string.format("%s/%s", folderName, originalPath)

        if isModuleAvailable(newPath) then
            return oldRequire(newPath)
        end
    end

    return oldRequire(originalPath)
end
```

We found out that, when requiring our [Class](https://github.com/yfrit/yfritlib/blob/master/Class.lua) module, instead of receiving a table when doing `local class = require("YfritLib.Class")`, we were receiving a string `Lua/YfritLib/Class.lua`. We tracked this down to happening only when `pcall` function calls were triggered inside of coroutines. We also found out that everything ran fine when the package loading happened on the main thread, as the coroutine require would use the cached table. With that in mind, this was our workaround:

```lua
local oldRequire = _G.require

--create a loader that tries to require a Lua file everytime something is required
local function isModuleAvailable(name)
    local _, isMainThread = coroutine.running()
    if not isMainThread then
        -- require outside coroutine first because things bug in here
        coroutine.yield(
            function(callback)
                pcall(oldRequire, name)
                callback()
            end
        )
    end

    local ok, errorMessage = pcall(oldRequire, name)
    if ok or not errorMessage:match(string.format("module '%s' not found", name)) then
        return true
    else
        return false
    end
end

--"Lua"
local attempts = {
    "Lua",
    "Lua/CardGameLogic",
    "Lua/MapExplorerLogic"
}

function _G.require(originalPath)
    originalPath = originalPath:gsub("%.", "/")

    for _, folderName in ipairs(attempts) do
        --"Module" => "Lua/Module" or "Lua/CardGameLogic/Module" or "Lua/MapExplorerLogic/Module"
        local newPath = string.format("%s/%s", folderName, originalPath)

        if isModuleAvailable(newPath) then
            local result = oldRequire(newPath)

            if type(result) == "string" then
                print("[DEBUG]", debug.traceback())
            end

            return result
        end
    end

    return oldRequire(originalPath)
end
```

By using a callback, we are able to make the `pcall` call from outside the coroutine, caching the result. With this, our project now runs completely fine. I wanted to describe everything in detail is might be helpful for you and the plugin, but once again I want to make it clear we have a lot of cases like the `_G.require` override you see in this example, which means it's completely possible (I'd even say likely) that our code is the cause of this strange behaviour.

---

Finally, this reply is already way too big, we ran into a different problem after fixing the require issues. This is a completely unrelated problem, so I can also open a different issue if you think it's better. We were using a JSON file as a static database file. The file was loaded by getting the Content Directory path from Unreal and passing it to the Lua scripts. This worked fine when the `Lua` directory was physically located within the Content folder, which is not true now for the packaged game (the Lua files are packed into the Pak file), so we were unable to find a way of accessing the JSON file. As a workaround, we transformed the JSON file into a Lua table, which works fine in practice, but it's not as pretty. Is there a way to actually do File I/O operations on non-lua files that are packed?

---

Alright, that's all! Sorry for the big wall of text, I hope this is helpful to you and the project. In the end we were just left with these two issues: the cooking error and the JSON IO problem. Since we have workarounds for both, they're not too pressing, but it would be really nice to have a properly built LuaMachine plugin hahaha. Thanks once again for your patience and all the help. Congratulations again on the incredible plugin. If you want me to open separate issues for these, just let me know.
